paddle/fluid/operators/mkldnn/fc_mkldnn_op.cc

```diff
@@ -16,384 +16,126 @@ limitations under the License. */
-    return dst_mem;
-  }
-
-  template <typename T>
-  static mkldnn::memory::desc CreateMemDescriptor(
-      const std::vector<int64_t>& dims, MKLDNNMemoryFormat format) {
-    return platform::MKLDNNMemDesc(dims, platform::MKLDNNGetDataType<T>(),
-                                   format);
-  }
-
-  template <typename T>
-  static mkldnn::memory::desc CreateMemDescriptor(const Tensor* tensor,
-                                                  MKLDNNMemoryFormat format) {
-    auto dims = framework::vectorize(tensor->dims());
-    return CreateMemDescriptor<T>(dims, format);
-  }
-
-  template <typename T>
-  mkldnn::memory CreateMemory(const mkldnn::memory::desc& desc,
-                              const Tensor* tensor) {
-    return CreateMemory(desc, platform::to_void_cast<T>(tensor->data<T>()));
-  }
-
-  mkldnn::memory CreateMemory(const mkldnn::memory::desc& desc, void* data) {
-    return memory(desc, engine_, data);
-  }
-
-  template <typename T>
-  std::shared_ptr<mkldnn::memory> CreateMemoryToBeCached(
-      const mkldnn::memory::desc& desc, const Tensor* tensor) {
-    return CreateMemoryToBeCached(desc,
-                                  platform::to_void_cast<T>(tensor->data<T>()));
-  }
-
-  std::shared_ptr<mkldnn::memory> CreateMemoryToBeCached(
-      const mkldnn::memory::desc& desc, void* data) {
-    return std::make_shared<memory>(desc, engine_, data);
-  }
-
-  // Create weights memory and transform to default MKL-DNN format
-  std::shared_ptr<mkldnn::memory> CreateWeightsMemory(const Tensor* weights) {
-    auto dims = framework::vectorize(weights->dims());
-    std::swap(dims[0], dims[1]);  // Correct output dimensions
-    auto src_desc = CreateMemDescriptor<float>(dims, MKLDNNMemoryFormat::io);
-    auto dst_desc = CreateMemDescriptor<float>(dims, MKLDNNMemoryFormat::oi);
-    // Transpose weights through MKL-DNN's reorder from io to oi format.
-    return Reorder(src_desc, dst_desc,
-                   platform::to_void_cast<float>(weights->data<float>()));
-  }
-
-  void CacheWeightsAndBias(const MKLDNNDeviceContext& dev_ctx,
-                           const ExecutionContext& ctx) {
-    std::string key = platform::CreateKey(dev_ctx);
-    key = platform::ExtendKeyWithThreadInfoIfNeeded(dev_ctx, key);
-
-    const std::string weights_key = key + ctx.InputName("W");
-    const std::string bias_key = key + ctx.InputName("Bias");
-    dev_ctx.SetBlob(weights_key, weights_);
-    dev_ctx.SetBlob(bias_key, bias_);
   }
 
   // Compute the bias scales so that its values correspond to the
   // scale of data being an output of weights and input multiplication
-  std::vector<float> ComputeBiasScales(const ExecutionContext& ctx) {
-    auto scale_in_data = ctx.Attr<float>("Scale_in");
-    auto scale_weights_data = ctx.Attr<std::vector<float>>("Scale_weights");
-    const size_t weight_scales_num = scale_weights_data.size();
-    std::vector<float> bias_scales(weight_scales_num);
+  std::vector<float> ComputeBiasScales(
```

Comment:
I would suggest caching scales of bias to avoid computing them every iteration. See corressponding code in: https://github.com/PaddlePaddle/Paddle/pull/37079/files (get_bias_scales) . Once you add scales to cache please drop #omp parallel to avoid spawning threads